### ISSUE TITLE
CSP problem: removing eval restriction

### ISSUE THREAD

**Issue by li-zyang** [Tampermonkey/tampermonkey#1024]
## Expected Behavior
Scripts runs on websites whose CSP header disable eval()

## Actual Behavior
They don't.
Console output: 

```
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”). 11 eval:3:89

Syntax error @ "scnu-easier-login"!
##########################
JSHINT output:
##########################

F_a/p</<@eval:3:90
c@eval:2:142
E_u/<@eval:3:259
ka@eval line 1 > Function:62:78
create@eval line 1 > Function:73:450
d@eval line 1 > Function:13:89
runListeners@eval line 1 > Function:13:52
anonymous/</</<@eval line 1 > Function:75:4
F@eval:11:116
eval line 1 > Function:63:20

Uncaught EvalError: call to eval() blocked by CSP
    p eval:3
    c eval:2
    E_u eval:3
    ka eval line 1 > Function:62
    create eval line 1 > Function:73
    d eval line 1 > Function:13
    runListeners eval line 1 > Function:13
    anonymous eval line 1 > Function:75
    F eval:11
eval:3:90
```

## Specifications

- Firefox 79.0 (64-bit) (latest)
- TM: 4.11.6118 (BETA)
- OS: Windows 10 1803

## Script

```js
// ==UserScript==
// @name         scnu-easier-login
// @namespace    https://github.com/li-zyang/
// @version      0.1.0
// @description  Automatically complete the fields on sso platform of scnu
// @author       阿昭
// @include      https://sso.scnu.edu.cn/*/login.html*
// @include      https://jwxt.scnu.edu.cn/xtgl/login_slogin.html*
// @require      https://cdn.staticfile.org/jquery/3.4.1/jquery.min.js
// @grant        GM_setValue
// @grant        GM_getValue
// @grant        GM_listValues
// @grant        GM_deleteValue
// @grant        GM_addValueChangeListener
// @grant        GM_removeValueChangeListener
// @grant        GM_addStyle
// @grant        GM_setClipboard
// @grant        GM_xmlhttpRequest
// @grant        GM_registerMenuCommand
// @grant        GM_unregisterMenuCommand
// @grant        GM_download
// @grant        GM_getTab
// @grant        GM_saveTab
// @grant        GM_getTabs
// @grant        GM_setClipboard
// @grant        GM_info
// @grant        GM_openInTab
// @grant        GM_notification
// @grant        GM_getResourceText
// @grant        GM_getResourceURL
// @connect      *
// @noframes
// @run-at       document-end
// @note         v1.0.0      1970-01-01  Add your release notes here.
// ==/UserScript==

(function(window, $) {

  'use strict';

  window.scnuEasierLoginJS = {};
  var s = scnuEasierLoginJS;

  var accounts = GM_getValue('accounts') || {};

  s.setAccount = function setAccount(name, field, value) {
    accounts[name] = accounts[name] || {};
    accounts[name][field] = value;
    GM_setValue('accounts', accounts);
  }
  var setAccount = s.setAccount;
  // account structure: { account, password }

  s.delAccount = function delAccount(name) {
    delete accounts[name];
    GM_setValue('accounts', accounts);
  }
  var delAccount = s.delAccount;

})
```

## Note
I've set `Modify existing content security policy (CSP) headers` to `Remove entirely (possibly unsecure)`

**Comment by martixy** [comment]
How does firefox deal with service workers?

I have encountered a similar issue in chrome. Tampermonkey can modify or strip CSP from network responses, but not from service worker responses.

The solution (it's more of a workaround) for me was to enable an option in chrome dev tools -> Service workers to bypass the service worker and always fetch from network.

**Comment by li-zyang** [comment]
> 
> 
> How does firefox deal with service workers?
> 
> I have encountered a similar issue in chrome. Tampermonkey can modify or strip CSP from network responses, but not from service worker responses.
> 
> The solution (it's more of a workaround) for me was to enable an option in chrome dev tools -> Service workers to bypass the service worker and always fetch from network.

Well...I haven't tested it out but bypassing service worker seems to degrade the experience when the computer is offline on some site...

This issue probably have nothing to do with the service worker because the site where this error occurs does not use a service worker (I can't find anyone relating to that site in the extension debugger)

**Comment by martixy** [comment]
You know what service workers are, right? Of course bypassing SW will degrade the experience.

In chrome at least, extensions don't have access to service worker responses, therefore can't modify CSP when service worker is used. I don't know what you mean by "extension debugger", but a site's service worker is located in the **Application** tab in FF devtools.

**Comment by raszpl** [comment]
How about hot patching service workers?

```
const origWorker = window.Worker
window.Worker = class Worker extends origWorker {
  constructor(script) {
    var req = new XMLHttpRequest();
    req.open('GET', script, false);
    req.send();
    var newBlobStr = `

---whatever you want to override goes here---

${ req.responseText }
`
      super(URL.createObjectURL(new Blob([newBlobStr])));
  }
}
```

Im not sure about fetch api header access. I use something like this to message service worker fetches back to main thread, GM_xmlhttpRequest it and pump it back into sw.

**Comment by derjanb** [comment]
This issue should be fixed at TM BETA 4.12.6126. 😅

I'm not sure how long it'll take for the Chrome Webstore team to review the new version, but in the meantime, you can download it from [here](https://www.tampermonkey.net/crx/tampermonkey_beta_4_12_6126.crx).
